helion/_compiler/matmul_utils.py

```diff
@@ -0,0 +1,299 @@
+        rhs_dtype: PyTorch dtype of the right operand
+        m: M dimension (rows of lhs) - can be symbolic or concrete value
+        n: N dimension (cols of rhs) - can be symbolic or concrete value
+        k: K dimension (cols of lhs/rows of rhs) - can be symbolic or concrete value
+        shape_str_fn: Optional function to generate shape strings from dimension lists
+        acc_dtype: Optional dtype to cast the result to (when not fusing accumulator)
+        input_precision: Optional precision for the dot operation (e.g., 'tf32')
+        out_dtype: Optional output dtype for the dot operation
+
+    Returns:
+        AST expression for the (potentially padded and unpadded) dot product result
+
+    Raises:
+        DimensionRequirementError: If K dimension is below minimum requirement (cannot be padded)
+
+    Notes:
+        - M and N dimensions can be padded if below hardware requirements
+        - K dimension cannot be padded and will raise an error if too small
+        - When padding is applied, the result is automatically unpadded using sum reductions
+    """
+    device = CompileEnvironment.current().device
+    min_m, min_n, min_k = min_dot_size(device, lhs_dtype, rhs_dtype)
+
+    # Try to get concrete values for comparisons
+    m_val = int(m) if isinstance(m, int) else None
+    n_val = int(n) if isinstance(n, int) else None
+    k_val = int(k) if isinstance(k, int) else None
+
+    pad_m = m_val is not None and m_val < min_m
+    pad_n = n_val is not None and n_val < min_n
+    pad_k = k_val is not None and k_val < min_k
+
+    if pad_k:
+        raise exc.DimensionRequirementError(
+            op="tl.dot",
+            dim="K",
+            min_val=min_k,
+            dtype1=lhs_dtype,
+            dtype2=rhs_dtype,
+            actual=k_val,
+        )
+
+    # Build kwargs
+    kwargs = {}
+    if input_precision is not None:
+        kwargs["input_precision"] = input_precision
+    if out_dtype is not None:
+        kwargs["out_dtype"] = out_dtype
+
+    # Fast path: no padding needed
+    if not (pad_m or pad_n):
+        if acc is not None:
+            kwargs["acc"] = acc
+        result = emit_tl_dot(lhs, rhs, **kwargs)
+        return cast_ast(result, acc_dtype) if acc_dtype else result
+
+    def _apply_reshape_mask_reshape(
+        tensor: ast.AST,
+        initial_shape: list[object] | str | None,
+        mask_expr: str,
+        final_shape: list[object] | str,
+    ) -> ast.AST:
+        """Apply the reshape/mask/reshape pattern for padding tensors.
+
+        This pattern is used to pad tensors to meet hardware requirements for matrix
+        multiplication. When a dimension is too small (e.g., M=3 but hardware needs M=16),
+        we need to pad it with zeros. The pattern works as follows:
+
+        1. Reshape: Expand the tensor to expose a new dimension. For example, to pad M=3
+           to M=16, we reshape from [3, K] to [3, ceil(16/3), K] = [3, 6, K].
```

Comment:
How does this reshape work?  Reshape requires the same number of elements before and after.

Also blocks must have power-of-2 sizes.  So 3 would not be legal.